c10/core/TensorImpl.cpp

```diff
@@ -184,6 +186,35 @@ TensorImpl::TensorImpl(
   // Caffe2 operators create Storages with default devices.
 }
 
+void TensorImpl::_change_backend_component_keys(c10::Device device) {
+  DispatchKey device_key =
+      computeDispatchKey(c10::nullopt, c10::nullopt, device);
+  BackendComponent new_backend =
+      c10::DispatchKeySet(device_key).highestBackendKey();
+  BackendComponent old_backend = key_set_.highestBackendKey();
+
+  // following logic TensorImpl::TensorImpl, update the BackendComponent related
+  // keys to correspond to device
+
+  auto key_set =
+      key_set_ - c10::getAutocastRelatedKeySetFromBackend(old_backend);
+  key_set = key_set | c10::getAutocastRelatedKeySetFromBackend(new_backend);
```

Comment:
I think these lines for autocast are necessary, but maybe add a TODO pointing out that Autocast should be a (per-backend) functionality key, and once we make that change then they'll no longer be necessary?